examples/keygen/create_primary.c

```diff
@@ -0,0 +1,248 @@
+                persistHandle = 0;
+            }
+        }
+
+        argc--;
+    }
+
+    XMEMSET(&primary, 0, sizeof(primary));
+    XMEMSET(&tpmSession, 0, sizeof(tpmSession));
+
+    printf("TPM2.0 Primary Key generation example\n");
+    printf("\tAlgorithm: %s\n", TPM2_GetAlgName(alg));
+    printf("\tUnique: %s\n", uniqueStr);
+    printf("\tAuth: %s\n", authStr);
+    printf("\tStore Handle: 0x%08x\n", persistHandle);
+    printf("\tUse Parameter Encryption: %s\n", TPM2_GetAlgName(paramEncAlg));
+
+    rc = wolfTPM2_Init(&dev, TPM2_IoCb, userCtx);
+    if (rc != TPM_RC_SUCCESS) {
+        printf("\nwolfTPM2_Init failed\n");
+        goto exit;
+    }
+
+    /* See if handle already exists */
+    if (persistHandle > 0) {
+        if (wolfTPM2_ReadPublicKey(&dev, &primary, persistHandle) == 0) {
+            printf("Primary Handle 0x%08x already exists\n", persistHandle);
+            goto exit;
+        }
+    }
+
+    /* Supported algorithms for primary key are only RSA 2048-bit & ECC P256 */
+    if (alg == TPM_ALG_RSA) {
+        rc = wolfTPM2_GetKeyTemplate_RSA_SRK(&publicTemplate);
+    }
+    else if (alg == TPM_ALG_ECC) {
+        rc = wolfTPM2_GetKeyTemplate_ECC_SRK(&publicTemplate);
+    }
+    else {
+        rc = BAD_FUNC_ARG;
+    }
+    if (rc != TPM_RC_SUCCESS) goto exit;
+
+    if (paramEncAlg != TPM_ALG_NULL) {
+        /* Start an authenticated session (salted / unbound) with parameter encryption */
+        rc = wolfTPM2_StartSession(&dev, &tpmSession, NULL, NULL,
+            TPM_SE_HMAC, paramEncAlg);
+        if (rc != TPM_RC_SUCCESS) goto exit;
+        printf("TPM2_StartAuthSession: sessionHandle 0x%x\n",
+            (word32)tpmSession.handle.hndl);
+
+        /* set session for authorization of the primary key */
+        rc = wolfTPM2_SetAuthSession(&dev, 1, &tpmSession,
+            (TPMA_SESSION_decrypt | TPMA_SESSION_encrypt | TPMA_SESSION_continueSession));
+        if (rc != TPM_RC_SUCCESS) goto exit;
+    }
+
+    /* optionally set a unique field */
+    if (uniqueStr != NULL) {
+        rc = wolfTPM2_SetKeyTemplate_Unique(&publicTemplate,
+            (const byte*)uniqueStr, (int)XSTRLEN(uniqueStr));
+        if (rc != TPM_RC_SUCCESS) goto exit;
+    }
+
+    printf("Creating new %s primary key...\n", TPM2_GetAlgName(alg));
+
+    rc = wolfTPM2_CreatePrimaryKey(&dev, &primary, hierarchy, &publicTemplate,
+        (const byte*)authStr, authStr ? (int)XSTRLEN(authStr) : 0);
+    if (rc != TPM_RC_SUCCESS) {
+        printf("wolfTPM2_CreatePrimaryKey failed\n");
+        goto exit;
+    }
+
+#ifdef DEBUG_WOLFTPM
+    printf("Primary Key Public (%d bytes)\n", primary.pub.size);
+    TPM2_PrintBin((const byte*)&primary.pub.publicArea, primary.pub.size);
+#endif
+
+    if (persistHandle > 0) {
+    #ifndef WOLFTPM_WINAPI
+        /* Move storage key into persistent NV */
+        printf("Storing Primary key to handle 0x%08x\n", persistHandle);
+        rc = wolfTPM2_NVStoreKey(&dev, hierarchy, &primary,
```

Comment:
Could the primary key be stored as persistent?

Reply:
> Could the primary key be stored as persistent?

Hi @JasonMac219 , on TPM's where the internal NV is available it is possible. On platforms like Windows TBS there is no way to store the primary key. You would need to recreate it using the same template and unique at startup which will create the same primary key, since it uses the same seed. Once it is created a temporary (memory based) handle is returned, which you can use until next reboot.